server/service/service_osquery.go

```diff
@@ -153,12 +156,30 @@ func (svc Service) EnrollAgent(ctx context.Context, enrollSecret, hostIdentifier
 		save = true
 	}
 	if save {
-		if err := svc.ds.SaveHost(ctx, host); err != nil {
-			return "", osqueryError{message: "saving host details: " + err.Error(), nodeInvalid: true}
+		if appConfig.ServerSettings.DeferredSaveHost {
+			go svc.serialSaveHost(host)
+		} else {
+			err = svc.ds.SaveHost(ctx, host)
+			if err != nil {
+				return "", errors.Wrap(err, "save host in enroll agent")
+			}
 		}
 	}
 
-	return host.NodeKey, nil
+	return nodeKey, nil
+}
+
+func (svc Service) serialSaveHost(host *fleet.Host) {
+	newVal := atomic.AddInt64(&counter, 1)
+	defer func() {
+		atomic.AddInt64(&counter, -1)
+	}()
+	level.Debug(svc.logger).Log("background", newVal)
+
+	err := svc.ds.SerialSaveHost(context.Background(), host)
```

Comment:
I get that this is because this (`serialSaveHost`, lowercase) is called in a goroutine, but this means we send a plain context down the stack, without any timeout, error handler (that is setup in the handler's context), logger, etc. Maybe that's fine as it will be logged here at least.

I guess we should set a reasonable maximal time for the goroutine to wait for the save to succeed - in pathological cases there could probably be another request to save the same host that could arrive while the previous one is still pending, and I don't think there are any guarantees which one would get through first.

Reply:
Q: If we do so, could it happen that `serialSaveHost` timeouts but the operation in `mysql` succeeds?

Reply:
In that case the host is saved but an error is logged, I don't think it is particularly problematic? Note that if it happens in a transaction and the context is expired by the time we try to commit, the commit will fail and it will rollback.

Reply:
Makes sense! Will add a deadline.